src/main/java/org/truffleruby/core/rope/ValidLeafRope.java

```diff
@@ -16,6 +16,8 @@
 
     public ValidLeafRope(byte[] bytes, Encoding encoding, int characterLength) {
         super(bytes, encoding, CodeRange.CR_VALID, encoding.isSingleByte(), characterLength);
+
+        assert !RopeOperations.isAsciiOnly(bytes, encoding) : "ASCII-only string incorrectly marked as CR_VALID";
```

Comment:
Should it also check it's not a broken CR?

Reply:
Probably. I was trying to balance the cost of the assertion. I'll add it and see if it's a problem.

Reply:
Thanks for pushing me on this. It caught an actual issue.

Reply:
We could also have `RopeOperations.isValidCodeRange()` so that only one scan is needed, but this is probably good enough.